ignite/metrics/gan/inception_score.py

```diff
@@ -0,0 +1,123 @@
+        device: specifies which device updates are accumulated on. Setting the metric's
+            device to be the same as your ``update`` arguments ensures the ``update`` method is non-blocking. By
+            default, CPU.
+
+    Example:
+
+    .. code-block:: python
+
+        from ignite.metrics.gan import InceptionScore
+
+        m = InceptionScore(splits=4)
+
+        y_pred = torch.rand((16,3,299,299))  # size must match the input shape of the InceptionV3 model
+
+        m.update(y_pred)
+
+        print(m.compute())
+
+    """
+
+    def __init__(
+        self,
+        splits: int = 10,
+        inception_model: Optional[nn.Module] = None,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+    ):
+        if inception_model is None:
+            try:
+                from torchvision import models  # type: ignore
+
+                inception_model = models.inception_v3(pretrained=True, transform_input=False)
+            except ImportError:
+                raise RuntimeError(
+                    "This metric requires torchvision to be installed. "
+                    "You may install torchvision using: \n pip install torchvision \n"
+                )
+        if not isinstance(inception_model, nn.Module):
+            raise TypeError(
+                f"Argument inception_model should be instance of nn.Module, but given {type(inception_model)}"
+            )
+        super(InceptionScore, self).__init__(output_transform=output_transform, device=device)
+        self.inception_model = inception_model.eval().to(self._device)
+        self.n_splits = splits
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._probs = []  # type: List[torch.Tensor]
+
+    @reinit__is_reduced
+    def update(self, output: torch.Tensor) -> None:
+        generated = output.detach()
+        inception_output = self.inception_model(generated)
+        probs = F.softmax(inception_output)
+        self._probs.append(probs)
+
+    def compute(self) -> float:
+        if len(self._probs) < 1:
+            raise NotComputableError("Inception score must have at least one example before it can be computed.")
+
+        ws = idist.get_world_size()
+        _probs_tensor = torch.cat(self._probs, dim=0)
+
+        if ws > 1 and not self._is_reduced:
+            _probs_tensor = cast(torch.Tensor, idist.all_gather(_probs_tensor))
+        self._is_reduced = True
+
+        result = 0.0
+        if idist.get_rank() == 0:
+            N = _probs_tensor.shape[0]
+            scores = torch.zeros((self.n_splits,))
+            for i in range(self.n_splits):
+                part = _probs_tensor[i * (N // self.n_splits) : (i + 1) * (N // self.n_splits)]
+                kl = part * (torch.log(part) - torch.log(torch.mean(part, dim=0)))
+                kl = torch.mean(torch.sum(kl, dim=1))
+                scores[i] = torch.exp(kl)
+            result = torch.mean(scores).item()
```

Comment:
Can't we accumulate `kl` on the `update` to avoid saving all `probs` ?

Reply:
If what you meant is this:
```python
    @reinit__is_reduced
    def update(self, output: torch.Tensor) -> None:
        generated = output.detach()
        with torch.no_grad():
            inception_output = self.inception_model(generated)
        probs = self._apply_softmax(inception_output)
        kl = self._compute_kl(probs)
        self._kl += kl
        self._num_examples += probs.shape[0]

    @sync_all_reduce("_kl","_num_examples")
    def compute(self) -> float:
        return self._kl.item() / self._num_examples
```
I think this is not valid. 

The problem is how to incorporate `splits` into the calculations , when you have the whole dataset it's easy to split it into well-defined segments ,calculate the `kl` for each and then average .

Reply:
@ahmedo42 you can think of `n_splits` of KL accumulators. Here is an examples with counter (which you can replace with KL) : 
```python
n_splits = 12
batch_size = 32
num_samples = 2000

n_iters = (num_samples + batch_size - 1) // batch_size 

# reset():
iteration = 0
counter = torch.zeros(n_splits)

for j in range(n_iters):
    
    # update():
    for i in range(batch_size):
        counter[iteration % n_splits] += 1
        iteration += 1

> 
> counter
> tensor([168., 168., 168., 168., 168., 168., 168., 168., 168., 168., 168., 168.])
```


Have you also seen this paper: https://arxiv.org/abs/1801.01973 where they suggest to get rid of `n_splits` ?

Reply:
@vfdev-5  thanks for the idea , this is the implementation based on what you said , the inception score from this is 10.3 (compared to 9.6 of the previous implementation under the same conditions):
```python

    @reinit__is_reduced
    def update(self, output: torch.Tensor) -> None:
        generated = output.detach()
        with torch.no_grad():
            inception_output = self.inception_model(generated)
        probs = self._apply_softmax(inception_output)
        batch_size = probs.shape[0]
        for i in range(batch_size):
            self._kl[self._iteration % self.n_splits] += self._compute_kl(probs[i])
            self._iteration += 1
        self._num_examples += probs.shape[0]

    @sync_all_reduce("_kl","_num_examples")
    def compute(self) -> float:
        return torch.mean(self._kl).item() / self._num_examples

    def _compute_kl(self,input_tensor:torch.Tensor):
        kl = input_tensor * (torch.log(input_tensor) - torch.log(torch.mean(input_tensor, dim=0)))
        kl = torch.sum(kl, dim=0)
        return torch.exp(kl)
```
I saw the paper and I think we should provide the original implementation ( as it's still widely used now ) , then we can extend that.

Reply:
I think this implementation is incorrect due to `_compute_kl` : 
- `torch.log(torch.mean(input_tensor, dim=0))` originally computes the average on `num_samples // n_splits` of tensors and in your case it is batch size
-  `torch.exp(kl)` should not be applied here.